drop unused: toplevel, assign-only

alexlamsl opened the pull request:
Only works when assignment is on the same scope level as declaration for now (see new tests in `drop-unused.js`).

Run against `angular.js` and managed to shave one extra variable off without any measureable change in performance.

This can obviously be extended in the future, but given how long I spent reading up `drop_unused()` let's take one step at a time. Putting it here as I run out of tests to throw at it (for now).

kzc commented:
How can your optimization be disabled for the `collapse_vars` tests because I don't want these variables to be dropped. This tests the flow analysis of collapse_vars.

Can your optimization be put behind a new option?

kzc commented on test/compress/drop-unused.js:
```diff
@@ -177,3 +177,58 @@ keep_fnames: {
+            var a;
+            a = 1;
+        }
+
```
Please remove the blank lines between the functions in this test so it's easier to compare the input to the output. Thanks.

kzc commented:
Perhaps a new option `drop_assign_only` defaulting to `true`? That way it can be selectively disabled for tests.

alexlamsl commented:
@kzc any suggestions with a shorter option name? It's a chore adding white spaces to all the other options... 😝

kzc commented:
Feel free to come up with a shorter name. `drop_unused` or `assign_only` maybe?

kzc commented:
Those suggested names are not the greatest. What's the longest option name currently?

alexlamsl commented:
`drop_debugger` and `collapse_vars` are tied.

kzc commented:
I'm stumped to come up with a better, shorter name.

alexlamsl commented:
@kzc no worries - I'll think of something after/while updating the code and tests 😉

kzc commented:
What if `unused` can be overloaded with an integer value `2` indicating to not use this optimization? If it's any other true-ish value then it was as it was before.

alexlamsl commented:
@kzc sounds pretty cool actually, which makes this new feature off by default as well.

kzc commented:
We can have the new feature enabled by default with `true`. 
Setting `unused=2` can disable the feature.

alexlamsl commented:
@kzc yes, I misread your previous comment (again)

kzc commented:
People tend use default options. Not many people set `collapse_vars=true`, for example.

> @kzc yes, I misread your previous comment (again)

You didn't misread it - I rewrote it. Sorry about that.

kzc commented:
What if there was yet another option setting that can remove unused top level vars?
```
$ echo 'var a = foo();' | bin/uglifyjs -c
var a=foo();
```
`var a` was not dropped because it was top-level. 

This should be off by default as some people expect this grandfathered behavior in uglified scripts in the wild.

kzc commented on test/compress/collapse_vars.js:
```diff
@@ -588,12 +587,11 @@ collapse_vars_lvalues: {
         function f2(x) { var z = x, a = ++z; return z += a; }
         function f3(x) { var a = (x -= 3); return x + a; }
         function f4(x) { var a = (x -= 3); return x + a; }
-        function f5(x) { var w = e1(), v = e2(), c = v = --x; return (w = x) - c; }
```
Can you please set `unused: 2` in the collapse_vars tests with different results?

alexlamsl replied:
I copied these original tests to the bottom with `unused=2` - guess I should do it the other way round 😅

kzc commented on test/compress/collapse_vars.js:
```diff
@@ -388,9 +388,8 @@ collapse_vars_do_while: {
-            var a = 2;
             do {
-                fn(a = 7);
+                fn(7);
```
Can you please set unused: 2 in the collapse_vars tests with different results?

kzc commented on test/compress/collapse_vars.js:
```diff
@@ -1209,3 +1207,42 @@ collapse_vars_regexp: {
     }
 }
+
+keep_assign: {
```
I see what you've done. Please keep the original test names in the original place. Just change `unused` to `2` in the original tests that have different results.

kzc replied:
You may add a new test below the original.

alexlamsl replied:
Would you like me to copy the whole section below each of the two original blocks of tests, or just the affected case?

kzc replied:
The latest collapse_vars test changes look good - thanks!

kzc commented on lib/compress.js:
```diff
@@ -1435,6 +1436,13 @@ merge(Compressor.prototype, {
+                    if (node instanceof AST_Assign && node.operator == "="
+                        && node.left instanceof AST_SymbolRef
+                        && scope === self
+                        && assign_as_unused) {
```
Would you mind making `assign_as_unused` the first condition in the `if`? It's faster if `assign_as_unused` is not in effect, and the same speed otherwise.

kzc commented on lib/compress.js:
```diff
@@ -1569,6 +1577,14 @@ merge(Compressor.prototype, {
                     }
+                    if (node instanceof AST_Assign && node.operator == "="
+                        && node.left instanceof AST_SymbolRef
+                        && assign_as_unused) {
```
Would you mind making assign_as_unused the first condition in the if? It's faster if assign_as_unused is not in effect, and the same speed otherwise.

alexlamsl commented on test/compress/collapse_vars.js:
```diff
@@ -1152,60 +1256,3 @@ collapse_vars_short_circuited_conditions: {
 }
-
-collapse_vars_regexp: {
-    options = {
```
Ooops!

kzc commented on test/compress/collapse_vars.js:
```diff
@@ -1152,60 +1256,3 @@ collapse_vars_short_circuited_conditions: {
     }
 }
-
-collapse_vars_regexp: {
```
Why was this regex test dropped?

alexlamsl replied:
PEBKAC

kzc replied:
This test was added originally as result of a bug fix. Just because the implementation with respect to `is_const` changed does not mean the test should be dropped.

kzc replied:
it's cool. i was just confused.

alexlamsl replied:
I was trying to remove my own tests at the end of the file, and forgot I've done so already 🙄 

Tests restored.

kzc commented:
There's a comment above that might have been lost in the shuffle: https://github.com/mishoo/UglifyJS2/pull/1450#issuecomment-275928244

Would welcome your thoughts on the best option to remove unused top level vars.

alexlamsl commented:
@kzc yeah I am still chewing on that thought - the code change [seems trivial](https://github.com/mishoo/UglifyJS2/blob/0610c020b1544820be9898a285ab6c9066490552/lib/compress.js#L1411) <sub>(famous last words...)</sub>

alexlamsl commented:
Recycling PR state to re-run timed-out CI tests.

kzc commented:
Dropping unused top-level vars may warrant its own option.

alexlamsl commented:
@kzc I've given it a go with string-based option value. Worse case, we can just break both of these features back out as separate flags.

I was hoping the code is the more challenging part of this exercise...

kzc commented:
I dunno - what do you think?

alexlamsl commented:
@kzc `options.compress.toplevel` sounds consistent with `options.mangle.toplevel` indeed, so I'll do just that.

I'll keep the string-based `keep_assign` as well.

alexlamsl commented:
@kzc any more thoughts on this?

Shall we keep this in `[WIP]` for a week while we come up with ways to break it?

kzc commented:
```
$ echo 'var a = foo();' | bin/uglifyjs -c toplevel
WARN: Side effects in initialization of unused variable a [-:1,4]
foo();
```
Perfect.

LGTM

kzc commented:
Feel free to drop [WIP] prefix and squash the commits.

kzc commented:
@alexlamsl Could you please add documentation of these new options in the README?

alexlamsl commented:
@kzc I knew I've forgotten _something_... 😅

kzc commented on README.md:
```diff
@@ -361,7 +361,11 @@ to set `true`; it's effectively a shortcut for `foo=true`).
   statically determine the condition
 
-- `unused` -- drop unreferenced functions and variables
+- `unused` -- drop unreferenced functions and variables (simplem direct variable
```
typo: simplem

kzc commented:
I didn't realize this was the case:
```
$ echo 'function f(x){}' | bin/uglifyjs -c toplevel
WARN: Dropping unused function f [-:1,9]
```
Can we drop top level functions under a *different* option than `toplevel`?  Otherwise many `uglify -c toplevel` minified files would be empty.

`toplevel`, in my opinion, should only apply to vars.

Is there a way to specify toplevel roots (by they vars of functions) to retain?

kzc commented:
Could we temporarily add [WIP] while this issue is sorted out?

kzc commented:
The more I think about it, `-c toplevel=true` probably *should* behave as you documented it, as it is consistent with how things are dropped in function scope.

But having a way to specify toplevel vars and functions to retain would be good though. Maybe something a `top_retain` option specifying a comma separated list of top level function/var names?

alexlamsl commented:
Sounds good to me either way - the latest commit just retain all top-level functions, but what's important in it is the logic which allows us to tell the two apart (which I think is working)

alexlamsl commented:
How about `toplevel` defaults to `false`, `true` removes everything (like before), then `"vars"`, `"funcs"` and `"vars,funcs"` for whatever is needed?

kzc commented:
What do you think about this? The proposed new `top_retain` options would contain either a comma separated list of names or a single regex for top level named things to keep:

i.e.:
```
-c top_retain=foo,bar,quux
```
where they're a mix of top level var and function names.

or
```
-c top_retain=/^MyApp_/
```
Mind you, these are *pre*mangle names.

The use of `top_retain` would imply that `-c toplevel=true`.

kzc commented:
> How about toplevel defaults to false, true removes everything (like before), then "vars", "funcs" and "vars,funcs" for whatever is needed?

That makes sense. I like it.

`top_retain` could still be implemented in addition to that for more granularity.

alexlamsl commented:
@kzc would be grateful if you can check this new `toplevel` behaviour for me while I implement `top_retain`

alexlamsl commented:
Implemented `top_retain`, also fixed a few corner cases with `toplevel` and `unused` - the short version is `AST_Scope.drop_unused()` was not coded with `AST_Toplevel` in mind, so added extra accounting for that.

(Re-open to trigger CI after time-out.)

kzc commented on README.md:
```diff
@@ -361,7 +361,15 @@ to set `true`; it's effectively a shortcut for `foo=true`).
+  assignments do not count as references unless set to `"keep_assign"`)
+
+- `toplevel` -- drop unreferenced functions (`"funcs"`) and/or variables (`"vars"`)
+  in the toplevel scope (disabled by default)
```
If `toplevel` is set to `true` then both unreferenced variables and functions will be dropped.

kzc commented on README.md:
```diff
@@ -361,7 +361,15 @@ to set `true`; it's effectively a shortcut for `foo=true`).
+
+- `top_retain` -- an array of names to prevent specific toplevel functions
+  and variables from removal even if unreferenced (has no effect when `toplevel`
+  is `false`)
```
> `top_retain` -- an array of names

`top_retain` -- a comma separated list of names ...

RegExp support for `top_retain` would be a nice to have.

> has no effect when `toplevel` is `false`

It would be nice if `top_retain` is set, then `toplevel` is implicitly set to `true`.

alexlamsl replied:
`top_retain` can also accept an array of strings, same way as `pure_funcs`

I guess I could feed all the top-level variables to the `RegExp` (or even a `function()`) and get it done that way.

Ah, yes, I forgot the implicit thing - makes sense since if you use `top_retain` you should have a good overview on the global variables by then.

kzc replied:
I've had no end of trouble with the pure_funcs CLI syntax: https://github.com/mishoo/UglifyJS2/issues/1265

I filed and closed two tickets on the thing because I keep forgetting the syntax.

kzc commented:
```
$ echo 'function y(){}' | bin/uglifyjs -c top_retain=y
undefined:2
return (y)
        ^

ReferenceError: y is not defined
```

kzc commented:
```
$ echo 'function y(){}' | bin/uglifyjs -c top_retain=y,toplevel
undefined:2
return (y)
        ^

ReferenceError: y is not defined
```

kzc commented:
I see the problem:
```
$ echo 'function y(){}' | bin/uglifyjs -c toplevel,top_retain='["x","y","z"]'
function y(){}
```
```
$ echo 'function y(){}' | bin/uglifyjs -c toplevel,top_retain='["x","z"]'
WARN: Dropping unused function y [-:1,9]
```
I'm not a big fan of that array syntax but I see why it might be required to not conflict with other compress options.

alexlamsl commented:
I actually tested using API so I didn't know there's breakage on command line. I am expecting `top_retain` to work with both comma-delimited string and array of strings.

I did manage to reproduce the error you had with `bin/uglifyjs` - will investigate.

kzc commented:
As I mentioned in https://github.com/mishoo/UglifyJS2/pull/1450#issuecomment-275953810 I think it's by design. A comma separated string cannot work on the CLI because yargs cannot distinguish between the comma separated compress options.

kzc commented:
So I guess the only outstanding things are `top_retain` RegExp support and implicitly setting `-c toplevel=true` when `top_retain` is used and `toplevel` is `undefined`.

alexlamsl commented:
@kzc ah so it's just a CLI limitation then. On that note I did have an urge to rewrite that in `commander` like I did for `html-minifier` 😛 

I've just finished updating the docs - will update code now and push in a moment.

kzc commented:
Ah, I see now - this is how one enters a string argument into yargs from the CLI:
```
$ echo 'function y(){}' | bin/uglifyjs -c top_retain='"x,y,z"',toplevel
function y(){}
```
```
$ echo 'function y(){}' | bin/uglifyjs -c top_retain='"x,z"',toplevel
WARN: Dropping unused function y [-:1,9]
```

alexlamsl commented:
So it's all a big `eval()`... ❗️ 

PTAL at the latest changes.

kzc commented:
The CLI and the various options work perfectly. Uglify users will definitely benefit from this new functionality.

Great work!

alexlamsl commented:
@kzc thanks for all the help, as always :+1:
